lib/gc/Transforms/MemRefToCPURuntime.cpp

```diff
@@ -0,0 +1,147 @@
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/SmallSet.h"
+
+namespace mlir {
+namespace gc {
+#define GEN_PASS_DEF_CONVERTMEMREFTOCPURUNTIME
+#include "gc/Transforms/Passes.h.inc"
+
+namespace {
+
+bool hasParallelParent(Operation *op) {
+  // Check if the parent contains a forall / parallel loop
+  for (Operation *parentOp = op->getParentOp(); parentOp != nullptr;
+       parentOp = parentOp->getParentOp()) {
+    if (parentOp->hasTrait<OpTrait::HasParallelRegion>()) {
+      return true;
+    }
+  }
+  return false;
+}
+struct AlignedAllocLowering : public OpRewritePattern<memref::AllocOp> {
+  using OpRewritePattern<memref::AllocOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(memref::AllocOp op,
+                                PatternRewriter &rewriter) const final {
+    auto loc = op->getLoc();
+    MemRefType type = op.getMemref().getType();
+    ValueRange symbolOperands = op.getSymbolOperands();
+    ValueRange dynamicSizes = op.getDynamicSizes();
+    cpuruntime::AllocOp newAllocOp = rewriter.create<cpuruntime::AllocOp>(
+        loc, type, dynamicSizes, symbolOperands);
+    if (hasParallelParent(op))
+      newAllocOp.setThreadLocal(true);
+    rewriter.replaceOp(op, newAllocOp.getResult());
+    return success();
+  }
+};
+
+struct AlignedDeallocLowering : public OpRewritePattern<memref::DeallocOp> {
+  using OpRewritePattern<memref::DeallocOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(memref::DeallocOp op,
+                                PatternRewriter &rewriter) const final {
+    auto loc = op->getLoc();
+    Value memref = op.getMemref();
+    rewriter.create<cpuruntime::DeallocOp>(loc, memref);
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+/// Given a memref value, return the "base" value by skipping over all
+/// ViewLikeOpInterface ops (if any) in the reverse use-def chain.
+static Value getViewBase(Value value) {
+  while (auto viewLikeOp = value.getDefiningOp<ViewLikeOpInterface>())
+    value = viewLikeOp.getViewSource();
+  return value;
+}
+
+struct ConvertMemRefToCPURuntime
+    : public impl::ConvertMemRefToCPURuntimeBase<ConvertMemRefToCPURuntime> {
+
+  void runOnOperation() final {
+    auto *ctx = &getContext();
+    // Create a local set to store operations that should not be transformed.
+    llvm::SmallSet<Operation *, 16> noTransformOps;
+
+    // Walk through the module to find func::FuncOp instances.
+    getOperation()->walk([&](func::FuncOp funcOp) {
+      BufferViewFlowAnalysis analysis(funcOp);
+      // Now walk through the operations within the func::FuncOp.
+      funcOp.walk([&](Operation *op) {
+        if (op->hasTrait<OpTrait::ReturnLike>()) {
+          for (Value operand : op->getOperands()) {
+            if (isa<MemRefType>(operand.getType())) {
+              Value v = getViewBase(operand);
```

Comment:
Do we still need this? I think alias analysis is enough.

Reply:
yes, alias analysis is enough, removed the `getViewBase`